### ISSUE TITLE
Losing Material Reference After Reimport Model

### ISSUE THREAD

**Issue by vitorbalbio** [godotengine/godot#42128]
**Godot version:**
Last 3.2.3 rc6 Bug also happen in all versions 3.2 and 4.0 versions 

**OS/device including version:**
Windows 10

**Issue description:**
After import and overwrite a model, Godot loses the Material in play but keeps working in editor until next reload of the scene. It's a destructible bug because once you reimport the model you lost the material assigned forever if it's not saved on the FileSystem.

-> Exported a gltf (or any other mesh format) that will overwrite some model already in the project.
-> Saved a external .mesh using the import options
-> Materials "Keep on Reimport" is check
-> The model is reimported without problems on the editor. It's update and materials are kept as correct.
-> But it breaks in play
-> Saving the scene and reload shows that the material was lost.

This breaks the scene setup every single time i need to change something in the original model. Also if you have the material set in the object without save it in the project it's lost forever.

![NVIDIA_Share_uvD98qbvKv](https://user-images.githubusercontent.com/6172013/93403797-df963980-f85e-11ea-8f34-d99fe57c959c.png)

**Steps to reproduce:**
Open the project. 
Open the .blend file in blender.
Change something in the model. Export and overwrite the Teste.glb file
It will update the current model in the Godot Editor
Play and see that the material (red) is lost.
Save and Reload.
The material is lost on the scene also

**Minimal reproduction project:**
[TesteBug.zip](https://github.com/godotengine/godot/files/5235478/TesteBug.zip)

**Comment by capnm** [comment]
I think the import/reload/cathig/etc. part of the editor is fundamentally broken from day 1 :(
It isn't just material see #26775
AFAIK the only workaround, that may in some cases help, is to restart Godot (and/or) delete the caches ...

**Comment by capnm** [comment]
>  Materials "Keep on Reimport" is check

I've now looked at your project. This is meant for the * .material file generated from the import. In the current Blender exporter you have to add some "fake" node-based material and then in the Godot double-click, change and save the generated * .material file in the inspector.
(Works for me, when different exported objects doesn't share same materials...).

**Comment by vitorbalbio** [comment]
> > Materials "Keep on Reimport" is check
> 
> I've now looked at your project. This option is meant for the ***.material** file generated from the import. In the current Blender exporter you have to add some "fake" node-based material and then in the Godot double-click, change and save the generated *.material file in the inspector.
> 
> (Works for me, when different exported objects don't use the same material – then the editor sometimes seems to overwrite the material file...).

I don't want to setup the material in blender. It's not even possible because i'm using custom shaders and need the material in the MeshInstance not in the Mesh. All that i want is that Godot stop breaking the setup and materials that i made in the MeshInstance. Shouldn't have any relation between the imported mesh (.mesh file) and something that you setup in the MeshInstance, It's just like reimport a texture and break your TextureRect or Sprite setup. This is just a reference to another resource. Don't know if it's a bug or a terrible design decision but i see as a critical bug.

**Comment by AndreaCatania** [comment]
After a debug turns out that this is not a bug, so let me try to explain what's going on here.
When you assign the material to the `Mesh` `Resource`, you are basically altering a `Resource` owned by the imported file (FBX,GLTF,COLLADA,...) and as you know we can't modify such resources.

So you have two ways to do it:
- Modify the material that the importer created for you.
- Assign a new material to the `MeshInstance`.

Regarding this issue, I'm leaving it open because we should add a **Warning** flag on the resource, that notify the user: "Any modification will be lost."
![Screenshot from 2020-11-20 14-47-49](https://user-images.githubusercontent.com/8342599/99807310-a2d90e00-2b3f-11eb-9545-f0fde5f13476.png)

**Comment by vitorbalbio** [comment]
Hi @AndreaCatania. Sorry but you totally misunderstood the problem. It's not about losing the material in the Mesh (it's totally normal since they are parte of the same resource) but in the MeshInstance that contains it. Check the first image I posted. I show the MeshInstance Material not the Mesh Material.

**Comment by AndreaCatania** [comment]
Reverted my changes, my apologies.

**Comment by vitorbalbio** [comment]
No problem. It would be great if someone could take a look on it anyway. 😊

**Comment by hoontee** [comment]
I can replicate this in the `3.2` branch but not in `master`. Can anyone else confirm that this is the case?

**Comment by Regnsloja** [comment]
Seconding, this behaviour is very hard to work with for me.
I have my MeshInstance, I assign some materials I want... Then when you alter your mesh file it reimports and now all those material assigns are gone and have to be put back. Again and again.
I feel like I'm missing something here, some setting that will make the reimport respect my MeshInstance changes.

Just pressing "Re-Import" doesn't seem to do this, only when you actually alter the mesh file.

I guess you're supposed to use the auto-generated imported materials, but I don't like to work with those.
I want to handle materials separately and manually, not connected to any imported files.

**Comment by Yazir** [comment]
I was getting way to annoyed having my MeshLibrary scene be flushed from materials every time I changed something, and no import setting worked (even not importing mats whatsoever) so I decided to do a workaround.

`tool
extends MeshInstance

export (Material) var force_material setget _set_force_material

func _init() -> void:
	if Engine.editor_hint and force_material:
		set("material/0", force_material)

func _set_force_material(mat : Material):
	force_material = mat
	if Engine.editor_hint and force_material:
		set("material/0", force_material)
`

Add it to MeshInstance which material you want to override. If you have any issues with it, try changing _init to _ready.

Enjoy!
